### DISCUSSION TITLE
Need help with env GitHub Action and check conditions

### DISCUSSION THREAD

**Discussion by calvertwong** [community/community#26000] [Actions]
<h1>Release APK</h1>
<pre><code>  - name: Build Signed APK
    run: ./gradlew assembleRelease
    
  - name: Sign Android release
    uses: r0adkll/sign-android-release@v1
    with:
      releaseDirectory: app/build/outputs/apk/release
      signingKeyBase64: ${{ secrets.SIGNING_KEY }}
      alias: ${{ secrets.ALIAS }}
      keyStorePassword: ${{ secrets.KEY_STORE_PASSWORD }}
      keyPassword: ${{ secrets.KEY_PASSWORD }}
      
  - name: Upload Release Artifact
    uses: actions/upload-artifact@v1
    with:
      name: app-release
      path: ${{ env.SIGNED_RELEASE_FILE }}
    env: 
      RELEASE_BUILD_SUCCESS: 'true' 
    
  - name: Send email indicate success
    if: ${DEBUG_BUILD_SUCCESS} == 'true' &amp;&amp; ${RELEASE_BUILD_SUCCESS} == 'true'
    uses: dawidd6/action-send-mail@v2
    with:
      server_address: smtp.gmail.com
      server_port: 465
      username: ${{secrets.MAIL_USERNAME}}
      password: ${{secrets.MAIL_PASSWORD}}
      subject: Github Actions job result
      body: Build job of ${{github.repository}} completed. Generated Debug and Release artifacts!
      to: ${{secrets.MAIL_USERNAME}}
      from: ${{secrets.MAIL_USERNAME}}
      
  - name: Send email indicate failure
    if: ${DEBUG_BUILD_SUCCESS} != 'true' || ${RELEASE_BUILD_SUCCESS} != 'true'
    uses: dawidd6/action-send-mail@v2
    with:
      server_address: smtp.gmail.com
      server_port: 465
      username: ${{secrets.MAIL_USERNAME}}
      password: ${{secrets.MAIL_PASSWORD}}
      subject: Github Actions job result
      body: Build job of ${{github.repository}} is incomplete. 
      to: ${{secrets.MAIL_USERNAME}}
      from: ${{secrets.MAIL_USERNAME}}
</code></pre>
<p>I just started messing around with GitHub Action and I am feeling stuck. I have been trying to achieve something like if(artifact exists) send Success email, else send Failure email but failed.</p>
<p>Does GitHub Action have if else statement at all? Can anyone look at my latest script and guide me how to check if an artifact exists and send success email, send failure email otherwise. Thanks in advanced.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="calvertwong" data-post="1" data-topic="119968">
<div class="title">
<div class="quote-controls"></div>
 calvertwong:</div>
<blockquote>
<p><code>    if: ${DEBUG_BUILD_SUCCESS} == 'true' &amp;&amp; ${RELEASE_BUILD_SUCCESS} == 'true'</code></p>
</blockquote>
</aside>
<p>First of, there are syntax issues here. The <code>if</code> conditional is not a shell, so you have to use the GitHub <code>${{ ... }}</code> syntax for variables, and reference environment variables through the <code>env</code> context:</p>
<pre><code class="lang-yaml">    if: ${{ env.DEBUG_BUILD_SUCCESS }} == 'true' &amp;&amp; ${{ env.RELEASE_BUILD_SUCCESS }} == 'true'
</code></pre>
<p>Secondly, I don’t see those environment variables getting set anywhere in the excerpt you posted. If they aren’t set in a way that isn’t obvious in the workflow you’ll have to do that. Depending on how you set them you might also want to use boolean literals (e.g. <code>true</code>) instead of strings (the current <code>'true'</code>) in the check.</p>
<p>Finally, there’s a <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions" rel="nofollow noopener">small note in the documentation</a> that <code>success()</code> will automatically be added to <code>if</code> conditionals that do not contain a status check function. Depending on what exactly you’re trying to do that may be relevant. :wink:</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Let me add something in addition.<br>
Firstly, when you use expressions in an  <code>if</code>  conditional, you may omit the expression syntax ( <code>${{ }}</code> ) because GitHub automatically evaluates the  <code>if</code>  conditional as an expression.<br>
You could use :</p>
<pre><code class="lang-auto">if: env.DEBUG_BUILD_SUCCESS == 'true' &amp;&amp; env.RELEASE_BUILD_SUCCESS == 'true'
</code></pre>
<p>Secondly, env context only include environment variables which is set in <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idenv" rel="nofollow noopener">Job level env</a> or <a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="nofollow noopener">set env command</a>. For <code>DEBUG_BUILD_SUCCESS</code>  and <code>RELEASE_BUILD_SUCCESS</code> , you need to set it to environment variable before you use it in if conditional.<br>
I am not sure how you confirm whether artifact exist or not. I you want to execute the send email steps based on the build step failed or succeeded, you could use <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions" rel="nofollow noopener">job status check function</a> in if conditional .</p>

**Comment by calvertwong** [comment]
<p>Thank you. I realized that the <code>if</code> syntax is wrong last night and made the changes.  I have used the below line to create an environment variable. In order to achieve my requirements, I had to rewrite the code logic and no longer checking for the existence of the artifact.</p>
<blockquote>
<p>run: echo ‘::set-env name=HAS_RELEASE_ARTIFACT::true’</p>
</blockquote>
